mutiny-core/src/fedimint.rs

```diff
@@ -0,0 +1,139 @@
+        self.archived.unwrap_or(false)
+    }
+}
+
+pub(crate) struct FedimintClient<S: MutinyStorage> {
+    pub _uuid: String,
+    pub child_index: u32,
+    pub federation_code: String,
+    pub fedimint_client: ClientArc,
+    stopped_components: Arc<RwLock<Vec<bool>>>,
+    storage: S,
+    network: Network,
+    pub(crate) logger: Arc<MutinyLogger>,
+    stop: Arc<AtomicBool>,
+}
+
+impl<S: MutinyStorage> FedimintClient<S> {
+    #[allow(clippy::too_many_arguments)]
+    pub(crate) async fn new(
+        uuid: String,
+        fedimint_index: &FedimintIndex,
+        federation_code: String,
+        xprivkey: ExtendedPrivKey,
+        storage: S,
+        network: Network,
+        logger: Arc<MutinyLogger>,
+    ) -> Result<Self, MutinyError> {
+        log_info!(logger, "initializing a new fedimint client: {uuid}");
+
+        // a list of components that need to be stopped and whether or not they are stopped
+        let stopped_components = Arc::new(RwLock::new(vec![]));
+
+        let stop = Arc::new(AtomicBool::new(false));
+
+        log_info!(logger, "Joining federation {}", federation_code);
+
+        let invite_code = InviteCode::from_str(&federation_code)
+            .map_err(|_| MutinyError::InvalidArgumentsError)?;
+
+        let mut client_builder = fedimint_client::Client::builder();
+        client_builder.with_module(WalletClientInit(None));
+        client_builder.with_module(MintClientInit);
+        client_builder.with_module(LightningClientInit);
+        client_builder.with_database(MemDatabase::new().into()); // TODO not in memory
+        client_builder.with_primary_module(1);
+        client_builder.with_federation_info(FederationInfo::from_invite_code(invite_code).await?);
+
+        let secret = create_fedimint_secret(xprivkey, fedimint_index)?;
+
+        let fedimint_client = client_builder.build(secret).await?;
+
+        Ok(FedimintClient {
+            _uuid: uuid,
+            child_index: fedimint_index.child_index,
+            federation_code,
+            fedimint_client,
+            stopped_components,
+            storage,
+            network,
+            logger,
+            stop,
+        })
+    }
+
+    pub fn fedimint_index(&self) -> FedimintIndex {
+        FedimintIndex {
+            child_index: self.child_index,
+            federation_code: self.federation_code.clone(),
+            archived: Some(false),
+        }
+    }
+}
+
+// A fedimint private key will be derived from `m/1'/X'`, where X is the index of a specific fedimint.
+// Fedimint will derive further keys from there.
+fn create_fedimint_secret(
+    xprivkey: ExtendedPrivKey,
+    fedimint_index: &FedimintIndex,
+) -> Result<DerivableSecret, MutinyError> {
+    let context = Secp256k1::new();
+    let xpriv = xprivkey.derive_priv(
+        &context,
+        &DerivationPath::from(vec![
+            ChildNumber::from_hardened_idx(1)?,
+            ChildNumber::from_hardened_idx(fedimint_index.child_index)?,
+        ]),
```

Comment:
we should incorporate the network into the secret, so mainnet is different than testnet

Reply:
Shouldnt we really be doing that at the highest level instead? We should want this everywhere, not just for fedimint.

Reply:
We could but really wouldn't be following the BIPs anymore, normally for derivation paths it is `m/Purpose/Network/Account/Internal/Index`

Reply:
Not going to be able to comply exactly with their derivations but the prefix essentially does once is goes into the fedimint client builder. I've also implemented network into this.